ZAPD/CrashHandler.cpp

```diff
@@ -0,0 +1,146 @@
+#include "CrashHandler.h"
+#include "Utils/StringHelper.h"
+
+#if defined(__linux__)
+void ErrorHandler(int sig)
+{
+	std::array<void*, 4096> arr;
+	constexpr size_t nMaxFrames = arr.size();
+	size_t size = backtrace(arr.data(), nMaxFrames);
+	char** symbols = backtrace_symbols(arr.data(), nMaxFrames);
+
+	fprintf(stderr, "\nZAPD crashed. (Signal: %i)\n", sig);
+
+	// Feel free to add more crash messages.
+	std::array<const char*, 3> crashEasterEgg = {
+		"\tYou've met with a terrible fate, haven't you?",
+		"\tSEA BEARS FOAM. SLEEP BEARS DREAMS. \n\tBOTH END IN THE SAME WAY: CRASSSH!",
+		"\tZAPD has fallen and cannot get up.",
+	};
+
+	srand(time(nullptr));
+	auto easterIndex = rand() % crashEasterEgg.size();
+
+	fprintf(stderr, "\n%s\n\n", crashEasterEgg[easterIndex]);
+
+	fprintf(stderr, "Traceback:\n");
+	for (size_t i = 1; i < size; i++)
+	{
+		Dl_info info;
+		uint32_t gotAddress = dladdr(arr[i], &info);
+		std::string functionName(symbols[i]);
+
+		if (gotAddress != 0 && info.dli_sname != nullptr)
+		{
+			int32_t status;
+			char* demangled = abi::__cxa_demangle(info.dli_sname, nullptr, nullptr, &status);
+			const char* nameFound = info.dli_sname;
+
+			if (status == 0)
+			{
+				nameFound = demangled;
+			}
+
+			functionName = StringHelper::Sprintf("%s (+0x%X)", nameFound,
+			                                     (char*)arr[i] - (char*)info.dli_saddr);
+			free(demangled);
+		}
+
+		fprintf(stderr, "%-3zd %s\n", i, functionName.c_str());
+	}
+
+	fprintf(stderr, "\n");
+
+	free(symbols);
+	exit(1);
+}
+#elif defined(_MSC_VER)
+
+void printStack(CONTEXT* ctx)
+{
+	BOOL result;
+	HANDLE process;
+	HANDLE thread;
+	HMODULE hModule;
+	ULONG frame;
+	DWORD64 displacement;
+	DWORD disp;
+
+#if defined(_M_AMD64)
+	STACKFRAME64 stack;
+	memset(&stack, 0, sizeof(STACKFRAME64));
+#else
+	STACKFRAME stack;
+	memset(&stack, 0, sizeof(STACKFRAME));
+#endif
+
+	char buffer[sizeof(SYMBOL_INFO) + MAX_SYM_NAME + sizeof(TCHAR)];
+	char module[512];
+
+	PSYMBOL_INFO symbol = (PSYMBOL_INFO)buffer;
+
+	CONTEXT ctx2;
+	memcpy(&ctx2, ctx, sizeof(CONTEXT));
+
+	process = GetCurrentProcess();
+	thread = GetCurrentThread();
+	SymInitialize(process, nullptr, TRUE);
+
+	constexpr DWORD machineType =
+#if defined(_M_AMD64)
+		IMAGE_FILE_MACHINE_AMD64;
+#else
+		IMAGE_FILE_MACHINE_I386;
+#endif
```

Comment:
Could you explain how this works? Is the expectation that we have Intel-specific and AMD-specific builds? I take it if one were to run AMD builds on an Intel CPU or vise versa the stack trace would not work?

Reply:
AMD64 is just 64bit and I386 is just 32 bit from what I can tell. If someone really wants ARM support its probably not too hard to add for them.

Reply:
`_M_X64` might make more sense. Apparently it is identical to `_M_AMD64`, but it won't confuse people into thinking their 64-bit Intel CPUs won't use it.

Reply:
You'd have a harder job getting Windows support for ARM, most likely.